# Laboratorio #7: FPGrowth Trees
___
Marcelo Delgado Mora C12510

Felipe Mena Rodríguez C24721

Jose Mora Monge C15114

___

## Imports

In [1]:
from mlxtend.frequent_patterns import fpgrowth, association_rules
from mlxtend.preprocessing import TransactionEncoder
import numpy as np
import pandas as pd
from google.colab import files
import time

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)


## **1. Dataset**

In [2]:
# Subir dataset:
archivo = files.upload()

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

Saving retail_data.csv to retail_data.csv


/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

In [3]:
# Obtener dataset:
transactions = pd.read_csv('retail_data.csv')
transactions.head()

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

Trans_ID Product1   Product2 Product3
0         1    Bread     Butter    Dairy
1         2   Nachos     Butter    Dairy
2         3    Juice        Jam      Egg
3         4    Juice        Jam      Egg
4         5    Juice  Vegetable    Salad

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

In [4]:
len(transactions)

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

2000

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

In [5]:
# Eliminar columnas con valores nulos:
transactions = transactions.dropna()
len(transactions)

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

2000

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)


In [6]:
transactions.drop(['Trans_ID'], axis=1, inplace=True)
transactions.head()

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

Product1   Product2 Product3
0    Bread     Butter    Dairy
1   Nachos     Butter    Dairy
2    Juice        Jam      Egg
3    Juice        Jam      Egg
4    Juice  Vegetable    Salad

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)


In [7]:
transactions['product_list'] = (
    transactions[['Product1', 'Product2', 'Product3']]
    .apply(lambda row: row.dropna().tolist(), axis=1)
)

print(transactions[['product_list']].head())

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

                product_list
0     [Bread, Butter, Dairy]
1    [Nachos, Butter, Dairy]
2          [Juice, Jam, Egg]
3          [Juice, Jam, Egg]
4  [Juice, Vegetable, Salad]


/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

## **2. Técnica para reglas de asociación: FP-Growth trees**

In [8]:
tiempo_fpgrowth_ini = time.perf_counter()
te = TransactionEncoder()
te_ary = te.fit_transform(transactions['product_list'].tolist())
encoded_transactions = pd.DataFrame(te_ary, columns=te.columns_, dtype=bool)

frequent_itemsets = fpgrowth(encoded_transactions, min_support=0.1, use_colnames=True)
association_rules_confidence = association_rules(frequent_itemsets, metric='confidence', min_threshold=0.5)
tiempo_fpgrowth_fin = time.perf_counter() - tiempo_fpgrowth_ini

frequent_itemsets = (
      frequent_itemsets
      .sort_values(by='support', ascending=False)
      .reset_index(drop=True)
)

confidence_rules = (
      association_rules_confidence
      .sort_values(by='confidence', ascending=False)
      .reset_index(drop=True)
)

frequent_itemsets['itemset_disp'] = (
    frequent_itemsets['itemsets'].apply(lambda s: ', '.join(sorted(s)))
)
confidence_rules['antecedents'] = confidence_rules['antecedents'].apply(lambda s: ', '.join(sorted(s)))
confidence_rules['consequents'] = confidence_rules['consequents'].apply(lambda s: ', '.join(sorted(s)))

print("\n-------------------Frequent itemsets----------------\n")
print(frequent_itemsets[['itemset_disp', 'support']].to_string(index=False))

print("\n----------------Association rules---------------\n")
print(confidence_rules[['antecedents', 'consequents', 'support', 'confidence']].to_string(index=False))


/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag


-------------------Frequent itemsets----------------

            itemset_disp  support
                     Egg   0.5760
                   Bread   0.3510
                     Jam   0.3410
                Egg, Jam   0.2570
              Bread, Egg   0.2505
                   Juice   0.2465
                  Butter   0.2465
                   Salad   0.2210
                  Nachos   0.2200
               Vegetable   0.2160
           Bread, Butter   0.2145
        Salad, Vegetable   0.2090
              Egg, Juice   0.2080
                   Dairy   0.2030
         Egg, Jam, Juice   0.2010
              Jam, Juice   0.2010
                   Salsa   0.1965
             Butter, Egg   0.1945
      Bread, Butter, Egg   0.1945
           Nachos, Salsa   0.1880
                  Fruits   0.1825
           Fruits, Salad   0.1825
Fruits, Salad, Vegetable   0.1705
       Fruits, Vegetable   0.1705
              Bread, Jam   0.1280
             Egg, Nachos   0.1175
              Egg, Salsa   

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

In [9]:
print("Tiempo ejecución Algoritmo FPGrowth: ", tiempo_fpgrowth_fin, " segundos")

Tiempo ejecución Algoritmo FPGrowth:  0.17908339400000273  segundos


/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

## **3. Selección de la medida: Kulczynski y Conviction**

In [10]:
association_rules_kulc = association_rules(frequent_itemsets, metric="kulczynski", min_threshold=0.6)

kulc_rules = (
    association_rules_kulc
    .sort_values(by='kulczynski', ascending=False)
    .reset_index(drop=True)
)


kulc_rules['antecedents'] = kulc_rules['antecedents'].apply(lambda s: ', '.join(sorted(s)))
kulc_rules['consequents'] = kulc_rules['consequents'].apply(lambda s: ', '.join(sorted(s)))

print("\n----------------Association rules (Kulczynski 0.6)---------------\n")
print(kulc_rules[['antecedents','consequents','support','kulczynski']].to_string(index=False))


association_rules_conv = association_rules(frequent_itemsets, metric="conviction", min_threshold=1.2)

conviction_rules = (
    association_rules_conv
    .sort_values(by='conviction', ascending=False)
    .reset_index(drop=True)
)

conviction_rules['antecedents'] = conviction_rules['antecedents'].apply(lambda s: ', '.join(sorted(s)))
conviction_rules['consequents'] = conviction_rules['consequents'].apply(lambda s: ', '.join(sorted(s)))

print("\n----------------Conviction rules (Conviction 1.2)---------------\n")
print(conviction_rules[['antecedents','consequents','support','conviction']].to_string(index=False))

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag


----------------Association rules (Kulczynski 0.6)---------------

      antecedents       consequents  support  kulczynski
            Salad         Vegetable   0.2090    0.956647
        Vegetable             Salad   0.2090    0.956647
            Salad            Fruits   0.1825    0.912896
           Fruits             Salad   0.1825    0.912896
            Salsa            Nachos   0.1880    0.905644
           Nachos             Salsa   0.1880    0.905644
            Salad Fruits, Vegetable   0.1705    0.885747
Fruits, Vegetable             Salad   0.1705    0.885747
 Salad, Vegetable            Fruits   0.1705    0.875018
           Fruits  Salad, Vegetable   0.1705    0.875018
        Vegetable            Fruits   0.1705    0.861799
           Fruits         Vegetable   0.1705    0.861799
        Vegetable     Fruits, Salad   0.1705    0.861799
    Fruits, Salad         Vegetable   0.1705    0.861799
            Salsa       Egg, Nachos   0.1175    0.798982
      Egg, Nachos   

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

## **4. Patrones Interesantes**

In [11]:
kulc_rules_subset = (
    kulc_rules
    .loc[(kulc_rules['kulczynski'] > 0.8)]
    .sort_values(['kulczynski'],
                 ascending=[False])
)

conf_rules_subset_kulc = (
    confidence_rules
    .loc[(confidence_rules['kulczynski'] <= 0.5)]
    .sort_values(['kulczynski'],
                 ascending=[False])
)

conv_rules_subset = (
    conviction_rules
    .loc[(conviction_rules['conviction'] > 10)]
    .sort_values(['conviction'],
                 ascending=[False])
)

conf_rules_subset_conv = (
    confidence_rules
    .loc[(confidence_rules['conviction'] <= 1.0)]
    .sort_values(['conviction'],
                 ascending=[False])
)

print("\n----------------Association rules Kulc > 0.8)---------------\n")
print(kulc_rules_subset[['antecedents','consequents','support', 'confidence', 'kulczynski']].to_string(index=False))

print("\n----------------Association rules Kulc (Filtered)---------------\n")
print(conf_rules_subset_kulc[['antecedents','consequents','support', 'confidence', 'kulczynski']].to_string(index=False))

print("\n----------------Association rules Conv > 10)---------------\n")
print(conv_rules_subset[['antecedents','consequents','support', 'confidence', 'conviction']].to_string(index=False))

print("\n----------------Association rules Conv (Filtered)---------------\n")
print(conf_rules_subset_conv[['antecedents','consequents','support', 'confidence', 'conviction']].to_string(index=False))


----------------Association rules Kulc > 0.8)---------------

      antecedents       consequents  support  confidence  kulczynski
            Salad         Vegetable   0.2090    0.945701    0.956647
        Vegetable             Salad   0.2090    0.967593    0.956647
            Salad            Fruits   0.1825    0.825792    0.912896
           Fruits             Salad   0.1825    1.000000    0.912896
            Salsa            Nachos   0.1880    0.956743    0.905644
           Nachos             Salsa   0.1880    0.854545    0.905644
            Salad Fruits, Vegetable   0.1705    0.771493    0.885747
Fruits, Vegetable             Salad   0.1705    1.000000    0.885747
 Salad, Vegetable            Fruits   0.1705    0.815789    0.875018
           Fruits  Salad, Vegetable   0.1705    0.934247    0.875018
        Vegetable            Fruits   0.1705    0.789352    0.861799
           Fruits         Vegetable   0.1705    0.934247    0.861799
        Vegetable     Fruits, Salad   0.

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

### **Kulczynski:**
Se encontraron reglas interesantes, como:
- Vegetable y Salad
- Fruits y Salad
- Nachos y Salsa

Donde sus coeficientes de Kulczynski son superiores a 0.9 (muy cerca de 1).
También se encontraron reglas con un coeficiente inferior pero igual significativo (>0.8):
- Salad y (Fruits, Vegetable)
- (Salad, Vegetable) y Fruits
- Vegetable y Fruits
- Vegetable y (Fruits, Salad)

Por otra parte, se puede apreciar ciertas reglas que tienen confianza y soporte alto, pero índice de Kulczynski bajo, por lo que podría ser un caso engañoso donde no existe una implicación real. Puede verse en:
- Dairy y Nachos, con confianza de 0.50 y Kulczynski de 0.48
- (Nachos, Salsa) y Egg, con confianza de 0.62 y Kulczynski de 0.41
- Salsa y Egg, con confianza de 0.59 y Kulczynski de 0.40
- Nachos y Egg, con confianza de 0.53 y Kulczynsky de 0.36





### **Conviction:**

Se encontraron reglas interesantes, como:
- Fruits y Salad
- (Fruits, Vegetable) y Salad
- (Egg, Salsa) y Nachos
- (Butter, Egg) y Bread
- (Jam, Juice) y Egg

Donde sus valores de Conviction son infinitos. Otras reglas con coeficiente significativo (>10):
- Vegetable y Salad
- (Egg, Juice) y Jam
- Salsa y Nachos
- Salad y Vegetable
- Fruits y (Salad, Vegetable)
- Fruits y Vegetable

Por otra parte, se puede apreciar una regla que tienen confianza y soporte alto, pero valor de Conviction bajo, por lo que podría ser un caso engañoso donde no existe implicación real. Esta es:

- Nachos y Egg, con confianza de 0.53 y Conviction de 0.91.
